rustls/src/msgs/handshake.rs

```diff
@@ -851,7 +851,7 @@ pub struct ClientHelloPayload {
     pub session_id: SessionID,
     pub cipher_suites: Vec<CipherSuite>,
     pub compression_methods: Vec<Compression>,
-    pub extensions: Vec<ClientExtension>,
+    pub extensions: Option<Vec<ClientExtension>>,
```

Comment:
This is a weird type to use because it allows encoding "No extensions" as `None` separately from "zero extensions" as `Some(vec![])`. Perhaps the invariant that the array is never empty is enforced somewhere when this is `Some` but IMO it would just be clearer to have one way of encoding "no extensions" as "zero extensions"; i.e. without the `Option` wrapper.

Reply:
The problem here is arising from the roundtripping requirement discussed in #603 -- "no extensions" and "zero extensions" have different and valid wire encodings (b"" and b"\x00\x00" respectively), so this needs to be expressed somewhere in the in-memory representation of ClientHelloPayload and ServerHelloPayload

I think I'm talking my way towards fixing #603 rather than doing this.